### PR TITLE
 [Name lookup] Use decl-based name lookup more regularly

### PR DESCRIPTION
**Pull request by DougGregor**
Avoid the use of the type checker in much of name lookup, relying on Decl-level information (not Type-level information) instead. There are several changes here that support each other to get back to a usable state:

* Use decl-based name lookup for unqualified lookup: replace uses of the `Type` form of `lookupQualified` (where the type was usually provided by `getSelfTypeInContext()`) with the "decl" form of the method, so we don't need to resolve types to perform unqualified name lookup.
* Use declaration-based name lookup in the conformance lookup table: similar to the above, allows us to determine the set of protocols to which a nominal type conforms using only Decl-level information.

With these changes, eliminate `LazyResolver::resolveExtensionForConformanceConstruction()` and some of the type-checker logic supporting it.

### CODE REVIEW

File: lib/AST/NameLookup.cpp [hunk 1]
```diff
@@ -2201,8 +2277,6 @@ directReferencesForTypeRepr(Evaluator &evaluator,
   }
 }
 
-/// Retrieve the set of type declarations that are directly referenced from
-/// the given type.
 static DirectlyReferencedTypeDecls
 directReferencesForType(Type type) {
```
Comment (slavapestov) [inline lib/AST/NameLookup.cpp]: This looks like it could fit on one line

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/AST/NameLookup.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/AST/NameLookup.cpp]: Ok. Will do in a follow-up

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/AST/NameLookup.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/AST/NameLookup.cpp]: It ended up in https://github.com/apple/swift/pull/18562/files

**Pull request merged**